capsules/src/rng.rs

```diff
@@ -0,0 +1,177 @@
+    buffer: Option<AppSlice<Shared, u8>>,
+    remaining: usize,
+    idx: usize,
+}
+
+impl Default for App {
+    fn default() -> App {
+        App {
+            callback: None,
+            buffer: None,
+            remaining: 0,
+            idx: 0,
+        }
+    }
+}
+
+pub struct SimpleRng<'a, RNG: rng::RNG + 'a> {
+    rng: &'a RNG,
+    apps: Container<App>,
+    getting_randomness: Cell<bool>,
+}
+
+impl<'a, RNG: rng::RNG> SimpleRng<'a, RNG> {
+    pub fn new(rng: &'a RNG, container: Container<App>) -> SimpleRng<'a, RNG> {
+        SimpleRng {
+            rng: rng,
+            apps: container,
+            getting_randomness: Cell::new(false),
+        }
+    }
+}
+
+impl<'a, RNG: rng::RNG> rng::Client for SimpleRng<'a, RNG> {
+    fn randomness_available(&self, randomness: &mut Iterator<Item = u32>) -> rng::Continue {
+        let mut done = true;
+
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                // Check if this app needs and random values.
+                if app.remaining > 0 && app.callback.is_some() && app.buffer.is_some() {
+                    app.buffer.take().map(|mut buffer| {
+                        // Check that the app is not asking for more than can
+                        // fit in the provided buffer.
+                        if buffer.len() < app.idx + app.remaining {
+                            app.remaining = buffer.len() - app.idx;
+                        }
+
+                        {
+                            // Add all available and requested randomness to the app buffer.
+
+                            // 1. Slice buffer to start from current idx
+                            let buf = &mut buffer.as_mut()[app.idx..];
+
+                            // 2. Take at most as many random samples as needed to fill the buffer
+                            //    (if app.remaining is not word-sized, take an extra one).
+                            let remaining_ints = if app.remaining % 4 == 0 {
+                                app.remaining / 4
+                            } else {
+                                app.remaining / 4 + 1
+                            };
+
+                            // 3. Zip over the randomness iterator and chunks
+                            //    of up to 4 bytes from the buffer.
+                            for (inp, outs) in randomness.take(remaining_ints)
+                                .zip(buf.chunks_mut(4)) {
+
+                                // 4. For each word of randomness input, update
+                                //    the remaining and idx and add to buffer.
+                                for (i, b) in outs.iter_mut().enumerate() {
+                                    *b = ((inp >> i * 8) & 0xff) as u8;
+                                    app.remaining -= 1;
+                                    app.idx += 1;
+                                }
+                            }
+                        }
```

Comment:
Well... the semantics of the language now are that variables live until the end of their scope. I think there was/is some talk about making the borrow-checker smarter but :man_shrugging: